mm2src/coins/lightning/ln_filesystem_persister.rs

```diff
@@ -96,43 +72,118 @@ impl LightningFilesystemPersister {
         path
     }
 
+    pub fn trusted_nodes_path(&self) -> PathBuf {
+        let mut path = self.main_path();
+        path.push("trusted_nodes");
+        path
+    }
+
     pub fn manager_path(&self) -> PathBuf {
         let mut path = self.main_path();
         path.push("manager");
         path
     }
-}
 
-#[derive(Clone)]
-pub struct LightningPersisterShared(pub Arc<LightningFilesystemPersister>);
+    pub fn monitors_path(&self) -> PathBuf {
+        let mut path = self.main_path();
+        path.push("monitors");
+        path
+    }
 
-impl Deref for LightningPersisterShared {
-    type Target = LightningFilesystemPersister;
-    fn deref(&self) -> &LightningFilesystemPersister { self.0.deref() }
-}
+    pub fn monitors_backup_path(&self) -> Option<PathBuf> {
+        if let Some(mut backup_path) = self.backup_path() {
+            backup_path.push("monitors");
+            return Some(backup_path);
+        }
+        None
+    }
 
-impl Persister<InMemorySigner, Arc<ChainMonitor>, Arc<Platform>, Arc<KeysManager>, Arc<Platform>, Arc<LogState>>
-    for LightningPersisterShared
-{
-    fn persist_manager(&self, channel_manager: &ChannelManager) -> Result<(), std::io::Error> {
-        FilesystemPersister::persist_manager(self.0.main_path().display().to_string(), channel_manager)?;
-        if let Some(backup_path) = self.0.backup_path() {
-            let file = fs::OpenOptions::new()
-                .create(true)
-                .write(true)
-                .truncate(true)
-                .open(backup_path)?;
-            channel_manager.write(&mut BufWriter::new(file))?;
+    /// Read `ChannelMonitor`s from disk.
+    pub fn read_channelmonitors<Signer: Sign, K: Deref>(
+        &self,
+        keys_manager: K,
+    ) -> Result<Vec<(BlockHash, ChannelMonitor<Signer>)>, std::io::Error>
+    where
+        K::Target: KeysInterface<Signer = Signer> + Sized,
+    {
+        let path = self.monitors_path();
+        if !path.exists() {
+            return Ok(Vec::new());
         }
-        Ok(())
+        let mut res = Vec::new();
+        for file_option in fs::read_dir(path).unwrap() {
+            let file = file_option.unwrap();
+            let owned_file_name = file.file_name();
+            let filename = owned_file_name.to_str();
+            if filename.is_some() && filename.unwrap() == "checkval" {
```

Comment:
I'm afraid that it's not safe to unwrap `OsString::to_str` since the path can contain non UTF-8 characters.

Reply:
Done